### ISSUE TITLE
"end" and "home" keys don't work with putty

### ISSUE THREAD

**Issue by dresc** [robbyrussell/oh-my-zsh#3757]
back to bash till this is fixed!!!

**Comment by skoftoby** [comment]
This is likely not an issue with oh-my-zsh. Have you verified that the problem does not occur without using oh-my-zsh?

If the issue is exclusive to zsh, does this help you? http://stackoverflow.com/questions/161676/home-end-keys-in-zsh-dont-work-with-putty

**Comment by jgrar** [comment]
Looks the same as #3733

**Comment by ncanceill** [comment]
**TL;DR** @dresc what makes you think your issue is related to OMZ?

***

As @skoftoby and @jgrar explained above, Zsh does not define Home and End keys, thus neither does OMZ. Consequently there is no obvious reason for this issue to be caused by OMZ.

Maybe you believe the issue is linked to a particular plugin, or a particular combination of plugins? a particular `~/.zshrc` setting?

Maybe the Home/End keys used to work, and you believe a certain update of OMZ may have broken them?

Maybe you tried disabling OMZ and the issue went away?

***

I also want to give you advice, but first I have a question.

> back to bash till this is fixed!!!

Is this supposed to be a joke? a threat?

I do not know what you imagine about OMZ maintainers, but we do not really care about "back to Bash" blackmail.

Now, here is advice about how to contribute to open-source software. When you report a problem, you should be answering three questions:
* how to reproduce the problem?
* what should happen?
* what actually happens?

Answering those three questions helps us figure out what is wrong and how to fix it.

We (usually) do not ask you to dive into the code and look for the problem yourself — we will do that for you. However, even before reporting the problem, we ask you to investigate and make sure that the issue is related to OMZ.

Please remember that **we are not payed to help you**. We simply do so benevolently because we love OMZ.

***

Sorry for the long reply. I believe the relevant question is at the top: what makes you think you issue is related to OMZ?

The morality here is: let us all be courteous and gentle to each other, this will only help making OMZ better.

**Comment by mcornella** [comment]
@dresc you need to be more graceful when asking for help. Please do me a favor and run `infocmp`, and post the output of that command between triple ticks \`\`\` so it's formatted nicely. 

FWIW, I do have *home* and *end* key support on my end. 
Cheers

**Comment by jgrar** [comment]
@dresc This might be a duplicate of #3061 could you join the discussion there and provide some further information?

Does this http://superuser.com/a/104016 help ?

**Comment by mcornella** [comment]
I'm closing this issue because it has become stale, and I already checked it on my end at https://github.com/robbyrussell/oh-my-zsh/issues/3757#issuecomment-91373916, so I don't have any more information to reproduce.

**Comment by apjanke** [comment]
Looks like this is specific to PuTTY. When I was [testing terminal key behaviors](https://docs.google.com/spreadsheets/d/1Ue5gb2hLjz058QEtEXwaw8YR7QoVZpSYzbXQKrRcGKk/edit?usp=sharing), I found PuTTY emitted different escape sequences for Home and End, which didn't match `xterm` or other emulators' behavior.

That SO question that @sometoby linked to is a good description of it and provides a workaround.

Before switching to `terminfo`-based key bindings, we used the local-cursor-mode escape sequence `^[[H` for Home, which is also what PuTTY produces, if you have it set to "Xterm R6 Function keys and keypad" in its configuration. But it does not change what it sends for Home and End in response to the `smkx` that switches in to cursor local mode, so the terminfo-based bindings won't work there. And its default behavior for Home/End doesn't match the character sequences issued for Home/End in any of the terminals I've seen. So, Home would have worked for some PuTTY users before, and stopped working after the change. Doesn't seem like End should have ever worked.

Sounds like a PuTTY issue. Though we could add workarounds for it, and other terminals, in OMZ.

**Comment by apjanke** [comment]
Ah, wait: looks like the Home and End key sequences generated by PuTTY by default (when "Function keys and keypad" are set to "ESC[n~") are those from the [DEC VT220 keyboard's "editing keypad"](http://vt100.net/docs/vt220-rm/chapter3.html). Also described in [the Xterm control sequences doc](http://invisible-island.net/xterm/ctlseqs/ctlseqs.html#h2-VT220-Style-Function-Keys). Putty is emitting VT220-style Home and End keys by default, and we're looking for PC Home/End keys in our key bindings.

```
                   Key        Normal     Application
                   ---------+----------+-------------
                   Insert   | CSI 2 ~  | CSI 2 ~
                   Delete   | CSI 3 ~  | CSI 3 ~
                   Home     | CSI 1 ~  | CSI 1 ~
                   End      | CSI 4 ~  | CSI 4 ~
                   PageUp   | CSI 5 ~  | CSI 5 ~
                   PageDown | CSI 6 ~  | CSI 6 ~
                   ---------+----------+-------------
```
